src/FuseGPUThreadLoops.cpp

```diff
@@ -293,6 +330,52 @@ class ExtractSharedAllocations : public IRMutator {
         }
     }
 
+
+    void visit(const ProducerConsumer *op) {
+        if (!in_threads) {
+            Stmt produce = mutate(op->produce);
+            if (!is_no_op(produce)) {
+                barrier_stage++;
+            }
+            Stmt update;
+            if (op->update.defined()) {
+                update = mutate(op->update);
+                if (!is_no_op(update)) {
+                    barrier_stage++;
+                }
+            }
+            Stmt consume = mutate(op->consume);
+
+            if (produce.same_as(op->produce) &&
+                update.same_as(op->update) &&
+                consume.same_as(op->consume)) {
+                stmt = op;
+            } else {
+                stmt = ProducerConsumer::make(op->name, produce, update, consume);
+            }
+        } else {
+            IRMutator::visit(op);
+        }
+    }
+
+    void visit(const Block *op) {
+        if (!in_threads && op->rest.defined()) {
```

Comment:
This looks like it's the same as the default visitor. Missing a barrier_stage++?